File: data-structure/tree/problems/invert_binary_tree.js
// https://leetcode.com/problems/invert-binary-tree/
//
//       4
//     /   \
//    2     7
//   / \   / \
//  1   3 6   9
//
//    becomes
//
//       4
//     /   \
//    7     2
//   / \   / \
//  9   6 3   1
//

function BT(v) {
    this.root = this.node(v);
}

BT.prototype.node = v => ({
    val: v,
    left: null,
    right: null
});

BT.prototype.insert = function (v) {
    const q = [this.root];

    while (q.length) {
        let node = q.shift();

        if (!node.left) return node.left = this.node(v);
        if (!node.right) return node.right = this.node(v);

        q.push(node.left);
        q.push(node.right);
    }
};

const tree = new BT(4);

[2,7,1,3,6,9].forEach(tree.insert.bind(tree));

console.log('tree.root', tree.root);

const invert = tree => {
    const q = [tree.root];
    const v = [];

    while (q.length) {
        const node = q.pop();

        const tmp = node.left;
        node.left = node.right;
        node.right = tmp;

        if (node.left) q.unshift(node.left);
        if (node.right) q.unshift(node.right);
    }
};

invert(tree);

console.log();

console.log('inverted tree.root', tree.root);

File: python/tree/invert_binary_tree.py
from Tree import Tree


def bfs(root):
    queue = [root]
    visited = []
    while queue:
        node = queue.pop()
        visited.append(node.value)
        if node.left:
            queue.insert(0, node.left)
        if node.right:
            queue.insert(0, node.right)
    return visited


# Iterative, return list.
def invert(root):
    queue = [root]
    visited = []

    while queue:
        node = queue.pop()
        visited.append(node.value)
        if node.left:
            queue.insert(0, node.left)
        if node.right:
            queue.insert(0, node.right)
    return visited


# Iterative, return new root.
def invert(root):
    queue = [root]
    while queue:
        node = queue.pop()
        node.right, node.left = node.left, node.right
        if node.left:
            queue.insert(0, node.left)
        if node.right:
            queue.insert(0, node.right)
    return root


# Recursive, return new root.
def invert(root):
    if not root:
        return

    root.right, root.left = root.left, root.right
    invert(root.left)
    invert(root.right)
    return root


nodes = [4, 2, 7, 1, 3, 6, 9]
# Inverted is: [4, 7, 2, 9, 6, 3, 1]
tree = Tree(nodes)
print(bfs(invert(tree.get_root())))
